README.md
Compute Degree Days
================

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/degday)](https://CRAN.R-project.org/package=degday)
[![degday status
badge](https://ajlyons.r-universe.dev/badges/degday)](https://ajlyons.r-universe.dev)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7083181.svg)](https://doi.org/10.5281/zenodo.7083181)
[![degree-day-challenge:
passing](https://raw.githubusercontent.com/ucanr-igis/degree-day-challenge/main/badges/degree-day-challenge-passing.svg)](https://ucanr-igis.github.io/degree-day-challenge/)
[![R-CMD-check](https://github.com/UCANR-IGIS/degday/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/UCANR-IGIS/degday/actions/workflows/R-CMD-check.yaml)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
<!-- badges: end -->

<a href='https://ucanr-igis.github.io/degday/'><img src='man/figures/logo.png' align="right" style="padding:15px; height:200px; width:177px;"/></a>**degday**
provides formulas for estimating [degree
days](https://en.wikipedia.org/wiki/Growing_degree-day) from daily
minimum and maximum temperatures. Degree days are commonly used in
agriculture to predict the development of plants and insects, which are
strongly correlated with accumulated heat above a certain a temperature.
To use the formulas, you must pass a record of daily minimum and maximum
temperatures, as well as a species-dependent temperature range in which
growth is possible.

Formulas are based on the work by Zalom et al ([1983](#references)) and
McMaster and Wilhelm ([1997](#references)). These same formulas are used
on the University of California [Integrated Pest
Management](http://ipm.ucanr.edu/WEATHER/) (UC IPM) website. See the UC
IPM website for additional background on [degree
days](http://ipm.ucanr.edu/WEATHER/ddconcepts.html), [detailed figures
and formulas](http://ipm.ucanr.edu/WEATHER/ddfigindex.html), and
[models](http://www.ipm.ucdavis.edu/MODELS/) that predict growth stages
of various crops and pests based on degree days.

`degday` implements all formulas from Zalom et al ([1983](#references)),
including the single and double triangle methods, and the single and
double sine methods. All formulas use the horizontal cutoff method.
`degday` does not currently support the vertical cutoff method, nor the
corrected sine and triangle methods. The simple average method is based
on McMaster and Wilhelm ([1997](#references)), which allows for an
optional upper threshold and supports the so-called ‘corn’ method for
dealing with temperatures below the lower and above the upper
thresholds.

You might be wondering, *which of the four estimation methods should I
use?* Most people compute degree days not as an end in itself, but
rather to look-up when a development milestone is predicted to take
place, such as blooming for a tree crop, or larvae emergence in an
insect. Hence you should use the same method that was used to build the
reference table. When in doubt, use the single-sine method. For more
info about the different methods, see Roltsch et al
([1999](#references)).

Degree days are sometimes referred to as *growing degree days*, which
generally refers to data that drives models of specifically plant
growth. Other terms that are more or less synonymous with degree days
are *heat units* and *thermal units*. *Chill hours* is a related
concept, but uses accumulated cold to predict plant and insect
development rather than accumulated heat. Formulas for computing chill
hours and chill portions may be found in
[chillR](https://cran.r-project.org/package=chillR).

## Installation

`degday` can be installed from CRAN:

    install.packages("degday")

The development version of `degday` is available on
<a href="https://ajlyons.r-universe.dev/" target="_blank"
rel="noopener">r-universe</a>. To install it, you can run:

    options(repos = c(ajlyons = 'https://ajlyons.r-universe.dev',
                      CRAN = 'https://cloud.r-project.org'))
    install.packages('degday')

Alternately, you can install the development version directly from
<a href="https://github.com/ucanr-igis/degday" target="_blank"
rel="noopener">GitHub</a>. (This requires the `remotes` package plus
<a href="https://cran.r-project.org/bin/windows/Rtools/" target="_blank"
rel="noopener">RTools</a> for Windows users.)

    remotes::install_github("ucanr-igis/degday")

## Example

To illustrate, we can compute degree days for a sample dataset
consisting of one year of minimum and maximum daily temperatures from
the
[Esparto.A](http://ipm.ucanr.edu/calludt.cgi/WXSTATIONDATA?MAP=yolo.html&STN=Esparto.A)
CIMIS weather station in Yolo County, California.

``` r
library(degday)
library(dplyr)

espartoa_csv <- system.file("extdata/espartoa-weather-2020.csv", package = "degday")
espartoa_temp <- read.csv(espartoa_csv) %>% mutate(date = as.Date(date))
espartoa_temp %>% head()
#>     station       date tmin tmax
#> 1 Esparto.A 2020-01-01   38   55
#> 2 Esparto.A 2020-01-02   36   67
#> 3 Esparto.A 2020-01-03   33   59
#> 4 Esparto.A 2020-01-04   37   59
#> 5 Esparto.A 2020-01-05   38   63
#> 6 Esparto.A 2020-01-06   36   58
```

To compute degree days, we have to tell it a range of temperatures in
which development occurs. We’ll select 55.0°F and 93.9°F, which are the
lower and upper thresholds for [Navel
Orangeworm](http://ipm.ucanr.edu/PHENOLOGY/ma-navel_orangeworm.html).

``` r
thresh_low <- 55
thresh_up <- 93.9
```

The single-triangle and single-sine methods can be computed with
`dd_sng_tri()` and `dd_sng_sine()`. We can add them as columns in our
table using `mutate()`:

``` r
espartoa_dd <- espartoa_temp %>%
  mutate(sng_tri = dd_sng_tri(daily_min = tmin, daily_max = tmax, 
                              thresh_low = thresh_low, thresh_up = thresh_up),
         sng_sine = dd_sng_sine(daily_min = tmin, daily_max = tmax, 
                                thresh_low = thresh_low, thresh_up = thresh_up))
#>  - using single triangle method
#>  - using single sine method

espartoa_dd %>% head()
#>     station       date tmin tmax sng_tri sng_sine
#> 1 Esparto.A 2020-01-01   38   55    0.00     0.00
#> 2 Esparto.A 2020-01-02   36   67    2.32     3.31
#> 3 Esparto.A 2020-01-03   33   59    0.31     0.68
#> 4 Esparto.A 2020-01-04   37   59    0.36     0.74
#> 5 Esparto.A 2020-01-05   38   63    1.28     1.99
#> 6 Esparto.A 2020-01-06   36   58    0.20     0.48
```

To compute degree days using the double-triangle and double-sine
methods, we need to first add a column to our temperature table for the
“day after” minimum temperature. That’s because these methods use the
minimum temperature of the next day to better model cooling in the
afternoon and evening hours.

We can add the next-day minimum temperature to our table with a little
dplyr.

``` r
espartoa_temp2 <- espartoa_temp %>%
  mutate(tmin_next = lead(tmin, n = 1))

espartoa_temp2 %>% head()
#>     station       date tmin tmax tmin_next
#> 1 Esparto.A 2020-01-01   38   55        36
#> 2 Esparto.A 2020-01-02   36   67        33
#> 3 Esparto.A 2020-01-03   33   59        37
#> 4 Esparto.A 2020-01-04   37   59        38
#> 5 Esparto.A 2020-01-05   38   63        36
#> 6 Esparto.A 2020-01-06   36   58        30
```

The double-triangle and double-sine methods can be computed with
`dd_dbl_tri()` and `dd_dbl_sine()`.

``` r
espartoa_dd2 <- espartoa_temp2 %>%
  mutate(dbl_tri = dd_dbl_tri(daily_min = tmin, daily_max = tmax, nextday_min = tmin_next,
                              thresh_low = thresh_low, thresh_up = thresh_up),
         dbl_sine = dd_dbl_sine(daily_min = tmin, daily_max = tmax, nextday_min = tmin_next,
                                thresh_low = thresh_low, thresh_up = thresh_up))
#>  - using double triangle method
#>  - using double sine method

espartoa_dd2 %>% head()
#>     station       date tmin tmax tmin_next dbl_tri dbl_sine
#> 1 Esparto.A 2020-01-01   38   55        36    0.00     0.00
#> 2 Esparto.A 2020-01-02   36   67        33    2.22     3.23
#> 3 Esparto.A 2020-01-03   33   59        37    0.34     0.71
#> 4 Esparto.A 2020-01-04   37   59        38    0.37     0.75
#> 5 Esparto.A 2020-01-05   38   63        36    1.23     1.95
#> 6 Esparto.A 2020-01-06   36   58        30    0.18     0.45
```

  

# References

Zalom, F.G., P.B. Goodell, L.T. Wilson, W.W. Barnett, and W.J. Bentley.
1983. *Degree-days: The calculation and use of heat units in pest
management*. UC DANR Leaflet 21373. Available from [Hathi
Trust](https://catalog.hathitrust.org/Record/008707238).

McMaster, G. S., and Wilhelm, W. W. 1997. *Growing degree-days: one
equation, two interpretations*. Agricultural and forest meteorology,
87(4), 291-300. Available from
[unl.edu](https://digitalcommons.unl.edu/cgi/viewcontent.cgi?article=1086&context=usdaarsfacpub)

Roltsch, W. J.; Zalom, F. G.; Strawn, A. J.; Strand, J. F.; Pitcairn, M.
J. 1999. *Evaluation of several degree-day estimation methods in
California climates*. Int. J. Biometeorol. 42:169-176.
<https://doi.org/10.1007/s004840050101>


NEWS.md
# degday 0.4.0 (2022-09-07) [![CRAN_Status_Badge](https://img.shields.io/badge/CRAN-published-blue.svg)](https://cran.r-project.org/package=degday)

* Initial release to CRAN
* `dd_calc()`: added `simp_avg_zero_method` argument for the so-called 'corn' variant of the simple average method
* `dd_simp_avg()`: new function (convenience wrapper for the simple average method)

# degday 0.3.0 (2022-08-21)

* `dd_fetch_cimis()` removed (will add to ca_weather package)

# degday 0.2.1 (2022-08-01)

* `dd_sng_tri()`, `dd_sng_sine()`, `dd_dbl_tri()`, `dd_dbl_sine()`: added `cutoff` argument
* `dd_sng_tri()`, `dd_sng_sine()`, `dd_dbl_tri()`, `dd_dbl_sine()`, `dd_calc()`: added `digits` argument for rounding; 
* `dd_dbl_tri()`, `dd_dbl_sine()`, `dd_calc()`: default value for `nextday_min` changed to `daily_min` (to match behavior on UC IPM website); default behavior when `nextday_min` is `NA` is to substitute `daily_min`

# degday 0.2.0 (2021-12-14)

* added `dd_fetch_cimis()`: fetches daily temperature data from weather stations in the CIMIS network (may move this to another package)
* `dd_calc()`: added `quiet` and `interpolate_na` arguments; updated to handle NA values

# degday 0.1.0

* Initial commit
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(degday)

test_check("degday")


tests/testthat/test-table7_tests.R
## This runs a series of tests using the verification table provided by Zalom et al 1983

## Load a copy of Table 7 from Zalom et al 1983, containing validation data
load("zalom83_tbl7.RData")
thresh_low <- 55
thresh_up <- 90

test_that("single triangle works", {
  my_gdd <- dd_calc(daily_min = tbl7_df$temp_min,
                     daily_max = tbl7_df$temp_max,
                     thresh_low = thresh_low,
                     thresh_up = thresh_up,
                     method = "sng_tri",
                     debug = FALSE)
  expect_equal(tbl7_df$sng_tri, my_gdd, tolerance = 0.001)
})

test_that("double triangle works", {
  my_gdd <- dd_calc(daily_min = tbl7_df$temp_min,
                     daily_max = tbl7_df$temp_max,
                     nextday_min = tbl7_df$temp_nextmin,
                     thresh_low = thresh_low,
                     thresh_up = thresh_up,
                     method = "dbl_tri",
                     debug = FALSE)
  ## We take out element #9 because I suspect the validation value in the table
  ## is actually an error
  expect_equal(tbl7_df$dbl_tri[-9], my_gdd[-9], tolerance = 0.001)
})

test_that("single sine works", {
  my_gdd <- dd_calc(daily_min = tbl7_df$temp_min,
                     daily_max = tbl7_df$temp_max,
                     thresh_low = thresh_low,
                     thresh_up = thresh_up,
                     method = "sng_sine",
                     debug = FALSE)
  expect_equal(tbl7_df$sng_sine, my_gdd, tolerance = 0.001)
})

test_that("double sine works", {
  my_gdd <- dd_calc(daily_min = tbl7_df$temp_min,
                     daily_max = tbl7_df$temp_max,
                     nextday_min = tbl7_df$temp_nextmin,
                     thresh_low = thresh_low,
                     thresh_up = thresh_up,
                     method = "dbl_sine",
                     debug = FALSE)
  expect_equal(tbl7_df$dbl_sine, my_gdd, tolerance = 0.001)
})










R/dd_calc.R
#' Estimate degree days from daily data
#'
#' Estimate degree days from daily min and max temperature
#'
#' @param daily_min Daily minimum temperature
#' @param daily_max Daily maximum temperature
#' @param nextday_min Minimum temp the day after
#' @param thresh_low Lower development threshold temperature
#' @param thresh_up Upper development threshold temperature
#' @param method Estimation method
#' @param cutoff Estimation cutoff method
#' @param digits Number of decimal places to round results to
#' @param cumulative Return cumulative values
#' @param no_neg Set negative values to zero
#' @param simp_avg_zero_method How to handle temperatures in the simple average method that fall outside the upper and lower thresholds (see details)
#' @param interpolate_na Interpolate missing values, logical
#' @param quiet Suppress messages, logical
#' @param debug Show additional messages
#'
#' @details Units for \code{daily_min}, \code{daily_max}, \code{thresh_low}, and \code{thresh_up} should all be the same
#' (i.e., all Fahrenheit or all Celsius). The function does not check for unit consistency.
#'
#' \code{nextday_min} is required for the double-triangle and the double-sine methods. These methods use the minimum temperature
#' of the following day to model temperatures in the 2nd half of the day. If omitted or NA, the daily minimum temperature will be
#' substituted.
#'
#' \code{no_neg = TRUE} sets negative values to zero. This is generally preferred when using degree days to predict the timing of
#'  development milestones, if one assumes that growth can not go backwards.
#'
#' The simple average method is taken from McMaster and Wilhelm (\href{https://digitalcommons.unl.edu/cgi/viewcontent.cgi?article=1086&context=usdaarsfacpub}{1997}). This method requires passing a lower threshold (also called the base temp). There are two ways of handling temperatures that fall below the base temperature. Most studies and applications use the default method (\code{simp_avg_zero_method = 1}) which simply 'zeroes out' average daily temperatures that fall below the base temp. Some studies (notably corn) use method 2, which truncates the daily minimum and maximum temperature before computing the simple average. Method 2 also allows you to pass an upper threshold. For details, see McMaster and Wilhelm.
#'
#' Missing values (NAs) in the temperatures will result in NA degree days. If \code{interpolate_na = TRUE}, missing degree days will
#' be interpolated. NAs in the middle of the series will be linearly interpolated, and NAs at the ends will be filled with the
#' adjacent values.
#'
#' @return A vector of estimated degree day values (either daily or cumulative, depending on the value of \code{cumulative})
#'
#' @examples
#' daily_temps <- system.file("extdata/espartoa-weather-2020.csv", package = "degday") %>%
#'   read.csv() %>%
#'     dplyr::mutate(date = as.Date(date)) %>%
#'     dplyr::slice(1:10)
#' daily_temps
#' ## Simple average method
#' dd_simp_avg(daily_min = daily_temps$tmin,
#'             daily_max = daily_temps$tmax,
#'             thresh_low = 55)
#' ## Single sine method
#' dd_sng_sine(daily_min = daily_temps$tmin, daily_max = daily_temps$tmax,
#'             thresh_low = 55, thresh_up = 93.9)
#' ## Single triangle method
#' dd_sng_tri(daily_min = daily_temps$tmin, daily_max = daily_temps$tmax,
#'            thresh_low = 55, thresh_up = 93.9)
#' ## Add next day min temp as an additional column
#' daily_temps_plus_tmin_next <- daily_temps %>% dplyr::mutate(tmin_next = dplyr::lead(tmin, n = 1))
#' daily_temps_plus_tmin_next
#' ## Double-triangle method
#' dd_dbl_tri(daily_min = daily_temps_plus_tmin_next$tmin,
#'            daily_max = daily_temps_plus_tmin_next$tmax,
#'            nextday_min = daily_temps_plus_tmin_next$tmin_next,
#'            thresh_low = 55, thresh_up = 93.9)
#'
#' @importFrom crayon yellow red
#' @importFrom zoo na.fill
#' @importFrom methods is
#' @export

dd_calc <- function(daily_min, daily_max, nextday_min = daily_min,
                     thresh_low = NULL, thresh_up = NULL,
                method = c("sng_tri", "dbl_tri", "sng_sine", "dbl_sine", "simp_avg")[0],
                cutoff = c("horizontal", "vertical", "intermediate")[1],
                digits = 2,
                cumulative = FALSE,
                no_neg = TRUE,
                simp_avg_zero_method = 1,
                interpolate_na = FALSE,
                quiet = FALSE,
                debug = FALSE) {

  if (!method %in% c("sng_tri", "dbl_tri", "sng_sine", "dbl_sine", "simp_avg")) stop("unknown value for `method`")
  if (!cutoff %in% c("horizontal", "vertical", "intermediate")) stop("unknown value for `cutoff`")
  if (cutoff != "horizontal") stop(paste0("The ", cutoff, " cutoff method is not yet supported"))
  if (length(daily_min) != length(daily_max)) stop("daily_min and daily_max must be the same length")
  if (FALSE %in% (daily_min <= daily_max)) stop("daily_min must always be less than daily_max")

  ## Get rid of the units. (alternatively I would need to check that thresh_low and thresh_up have the
  ## same units, otherwise some of the equality tests below will fail)
  if (is(daily_min, "units")) daily_min <- as.numeric(daily_min)
  if (is(daily_max, "units")) daily_max <- as.numeric(daily_max)

  ## Create a template error message
  err_missing_thresh <- "The {method} method requires you pass lower and upper threshhold. (If the parameters you were given do not include an upper threshold, pass a large number like 200.)"

  ## To handle NAs, find the indices where we have valid temps
  good_idx <- which(!is.na(daily_min) & !is.na(daily_max))

  if (length(good_idx) == 0) {
    stop("Invalid values for daily_min and/or daily_max (all NAs)")
  } else if (length(daily_min) != length(good_idx)) {
    if (!quiet) message(yellow(" - NAs found in daily_min and/or daily_max."))
    if (cumulative && !interpolate_na) stop(" - Can't return cumlative degree days because there are NAs in the input temperatures. You can let `interpolate_na = TRUE`, or fill the missing temperature values and try again.")
  }

  ## Initialize objects to hold results
  dd_all <- rep(NA, length(daily_min))  ## NAs will be replaced with computed dd at the very end
  degday <- NULL

  if (method == "simp_avg") {
    ### ------ SIMPLE AVERAGE -------
    #################################

    if (!quiet) message(yellow(" - using simple average method"))

    if (is.null(thresh_low)) stop(" - the simple average method requires a lower threshhold (base temp)")
    if (!simp_avg_zero_method %in% 1:2) stop("simp_avg_zero_method must be 1 or 2")
    if (!no_neg) stop("`no_neg` must be TRUE for the simple average method")

    thresh_low_use <- thresh_low

    ## If thresh_up is NULL, set a default value of 500
    thresh_up_use <- ifelse(is.null(thresh_up), 200, thresh_up)


    if (simp_avg_zero_method == 1) {
      ## First compute the average daily temp
      temp_avg <- (daily_max[good_idx] + daily_min[good_idx]) / 2

      ## If temp_avg falls outside thresh_low_use and thresh_up_use, manually crop it
      temp_avg[which(temp_avg < thresh_low_use)] <- thresh_low_use
      temp_avg[which(temp_avg > thresh_up_use)] <- thresh_up_use

      degday <- temp_avg - thresh_low_use

    } else if (simp_avg_zero_method == 2) {
      ## First coerce tmin and tmax to be within [thresh_low_use..thresh_up_use]

      ## We can work on daily_max rather than daily_max[good_idx] because which() throws out NAs
      daily_min[which(daily_min < thresh_low_use)] <- thresh_low_use
      daily_max[which(daily_max < thresh_low_use)] <- thresh_low_use

      ## tmax and tmin are both set to thresh_up if they are greater than thresh_low
      daily_min[which(daily_min > thresh_up_use)] <- thresh_up_use
      daily_max[which(daily_max > thresh_up_use)] <- thresh_up_use

      ## Now compute degree days
      degday <- ((daily_max[good_idx] + daily_min[good_idx]) / 2) - thresh_low_use

    }


  } else if (method == "sng_tri") {
    ### ------ SINGLE TRIANGLE -------
    ### Formulas come from Table 3 in Zalom et al 1983

    if (!quiet) message(yellow(" - using single triangle method"))
    if (is.null(thresh_low) || is.null(thresh_up)) {
      stop(red(sub("\\{method\\}", "single triangle", err_missing_thresh)))
    }

    for (i in good_idx) {
      this_case <- dd_case(daily_min[i], daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case ", this_case)))
      degday <- c(degday,
                  switch(this_case,
                         thresh_up - thresh_low, # case 1
                         0, # case 2
                         6 * (daily_max[i] + daily_min[i] - 2 * thresh_low) / 12, # case 3
                         (6 * (daily_max[i] - thresh_low) ^ 2 / (daily_max[i] - daily_min[i])) / 12, # case 4
                         (6 * (daily_max[i] + daily_min[i] - 2 * thresh_low) / 12) - ( ( 6 * (daily_max[i] - thresh_up) ^ 2 / (daily_max[i] - daily_min[i])) / 12  )  , # case 5
                         ((6 * (daily_max[i] - thresh_low)^2 / (daily_max[i] - daily_min[i]))  - (6 * (daily_max[i] - thresh_up)^2 / (daily_max[i] - daily_min[i]))) / 12) # case 6
                  )
    }


  } else if (method == "dbl_tri") {
    ### ------ DOUBLE TRIANGLE (HALF DAY FORMULAS) -------
    ### Formulas come from Table 4 in Zalom et al 1983

    if (!quiet) message(yellow(" - using double triangle method"))

    if (is.null(thresh_low) || is.null(thresh_up)) {
      stop(red(sub("\\{method\\}", "double triangle", err_missing_thresh)))
    }

    if (is.null(nextday_min)) {
      stop(red("The double triangle method requires you pass nextday_min"))
    }

    if (length(nextday_min) != length(daily_min)) stop("The next day minimum temperature must be the same length as the daily min and max temps")

    for (i in good_idx) {

      ## With the double-triangle half-day formulas, we need to
      ## compute two GDD values for each half day. This requires that
      ## the next day min temperature is passed.

      case_am <- dd_case(daily_min[i], daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case_am ", case_am)))
      dd_am <- dd_dbltri_half(case = case_am, tmin = daily_min[i], tmax = daily_max[i], thresh_low = thresh_low, thresh_up = thresh_up)

      nextday_min_use <- ifelse(is.na(nextday_min[i]), daily_min[i], nextday_min[i])
      case_pm <- dd_case(nextday_min_use, daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case_pm ", case_pm)))
      dd_pm <- dd_dbltri_half(case = case_pm, tmin = nextday_min_use, tmax = daily_max[i], thresh_low = thresh_low, thresh_up = thresh_up)

      degday <- c(degday, dd_am + dd_pm)

    }


  } else if (method == "sng_sine") {
    ### ------ SINGLE SINE -------
    ### Formulas come from Table 5 in Zalom et al 1983

    if (!quiet) message(yellow(" - using single sine method"))

    if (is.null(thresh_low) || is.null(thresh_up)) {
      stop(red(sub("\\{method\\}", "single sine", err_missing_thresh)))
    }

    for (i in good_idx) {

      this_case <- dd_case(daily_min[i], daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case = ", this_case)))

      alpha <- (daily_max[i] - daily_min[i]) / 2

      if (this_case == 1) {
        ss_dday <- thresh_up - thresh_low

      } else if (this_case == 2) {
        ss_dday <- 0

      } else if (this_case == 3) {
        ss_dday <- ((daily_max[i] + daily_min[i]) / 2) - thresh_low

      } else if (this_case == 4) {
        theta1 <- asin((thresh_low - (daily_max[i] + daily_min[i]) / 2) / alpha)
        ss_dday <- (1 / pi) * ((((daily_max[i] + daily_min[i]) / 2) - thresh_low) * (pi/2 - theta1) + alpha * cos(theta1))

      } else if (this_case == 5) {
        theta2 <- asin((thresh_up - (daily_max[i] + daily_min[i]) / 2) / alpha)
        ss_dday <- (1 / pi) * (( (( daily_max[i] + daily_min[i]) / 2) - thresh_low )*(pi/2 + theta2) + (thresh_up - thresh_low)*(pi/2 - theta2) - (alpha * cos(theta2)) )

      } else if (this_case == 6) {
        theta1 <- asin((thresh_low - (daily_max[i] + daily_min[i]) / 2) / alpha)
        theta2 <- asin((thresh_up - (daily_max[i] + daily_min[i]) / 2) / alpha)
        ss_dday <- (1 / pi) * ((((daily_max[i] + daily_min[i]) / 2) - thresh_low)*(theta2 - theta1) + alpha*(cos(theta1)-cos(theta2)) + (thresh_up - thresh_low)*(pi/2 - theta2))

      }

      degday <- c(degday, ss_dday)

    }


  } else if (method == "dbl_sine") {
    ### ------ DOUBLE SINE -------
    ### Formulas come from Table 6 in Zalom et al 1983

    if (!quiet) message(yellow(" - using double sine method"))

    if (is.null(thresh_low) || is.null(thresh_up)) {
      stop(red(sub("\\{method\\}", "double sine", err_missing_thresh)))
    }

    if (is.null(nextday_min)) {
      stop(red("The double sine method requires you pass nextday_min"))
    }

    if (length(nextday_min) != length(daily_min)) stop("The next day minimum temperature must be the same length as the daily min and max temps")

    for (i in good_idx) {

      ## With the double-sine half-day formulas, we need to compute two GDD values for each half day.
      case_am <- dd_case(daily_min[i], daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case_am ", case_am)))
      dd_am <- dd_dblsine_half(case = case_am, tmin = daily_min[i], tmax = daily_max[i], thresh_low = thresh_low, thresh_up = thresh_up)

      nextday_min_use <- ifelse(is.na(nextday_min[i]), daily_min[i], nextday_min[i])
      case_pm <- dd_case(nextday_min_use, daily_max[i], thresh_low, thresh_up)
      if (debug) message(yellow(paste0(" - ", i, ": case_pm ", case_pm)))
      dd_pm <- dd_dblsine_half(case = case_pm, tmin = nextday_min_use, tmax = daily_max[i], thresh_low = thresh_low, thresh_up = thresh_up)

      degday <- c(degday, dd_am + dd_pm)
    }

  } else {
    stop(red(paste0("Unknown method: ", method)))

  }

  ## Zero out negative values
  if (no_neg && sum(degday < 0) > 0) {
    if (!quiet) message(yellow(paste0(" - zeroing out ", sum(degday < 0), " negative degree day values")))
    degday[degday < 0] <- 0
  }

  ## Round values
  degday <- round(degday, digits)

  ## Swap in the computed values
  dd_all[good_idx] <- degday

  ## Interpolate missing values
  if (anyNA(dd_all) && interpolate_na) {
    if (!quiet) message(yellow(paste0(" - interpolating ", sum(is.na(dd_all)), " missing degree day values")))
    dd_all <- na.fill(dd_all, "extend")
  }

  ## Return degree days
  if (cumulative) {
    cumsum(dd_all)
  } else {
    dd_all
  }

}

#' @describeIn dd_calc Estimate degree days using the simple avg method
#' @export

dd_simp_avg <- function(daily_min, daily_max, thresh_low, thresh_up = NULL,
                        simp_avg_zero_method = 1, digits = 2, cumulative = FALSE,
                        quiet = FALSE) {

  dd_calc(daily_min = daily_min, daily_max = daily_max,
          thresh_low = thresh_low, thresh_up = thresh_up,
          method ="simp_avg", simp_avg_zero_method = simp_avg_zero_method,
          digits = digits, cumulative = cumulative, quiet = quiet)
}


#' @describeIn dd_calc Estimate degree days using the single-triangle method
#' @export

dd_sng_tri <- function(daily_min, daily_max, thresh_low = NULL, thresh_up = NULL,
                       cutoff = c("horizontal", "vertical", "intermediate")[1], digits = 2,
                       cumulative = FALSE, quiet = FALSE) {

  dd_calc(daily_min=daily_min, daily_max=daily_max, thresh_low=thresh_low, thresh_up=thresh_up,
          method ="sng_tri", cutoff=cutoff, digits=digits, cumulative=cumulative, quiet=quiet)
}

#' @describeIn dd_calc Estimate degree days using the single-sine method
#' @export

dd_sng_sine <- function(daily_min, daily_max, thresh_low = NULL, thresh_up = NULL,
                        cutoff = c("horizontal", "vertical", "intermediate")[1], digits = 2,
                        cumulative = FALSE, quiet = FALSE) {

  dd_calc(daily_min=daily_min, daily_max=daily_max, thresh_low=thresh_low, thresh_up=thresh_up,
          method ="sng_sine", cutoff=cutoff, digits=digits, cumulative=cumulative, quiet=quiet)
}

#' @describeIn dd_calc Estimate degree days using the double-triangle method
#' @export

dd_dbl_tri <- function(daily_min, daily_max, nextday_min = daily_min, thresh_low = NULL, thresh_up = NULL,
                       cutoff = c("horizontal", "vertical", "intermediate")[1], digits = 2,
                       cumulative = FALSE, quiet = FALSE) {

  dd_calc(daily_min=daily_min, daily_max=daily_max, nextday_min=nextday_min,
          thresh_low=thresh_low, thresh_up=thresh_up,
          method ="dbl_tri", cutoff=cutoff, digits=digits, cumulative=cumulative, quiet=quiet)
}

#' @describeIn dd_calc Estimate degree days using the double-sine method
#' @export

dd_dbl_sine <- function(daily_min, daily_max, nextday_min = daily_min, thresh_low = NULL, thresh_up = NULL,
                        cutoff = c("horizontal", "vertical", "intermediate")[1], digits = 2,
                        cumulative = FALSE, quiet = FALSE) {

  dd_calc(daily_min=daily_min, daily_max=daily_max, nextday_min=nextday_min,
          thresh_low=thresh_low, thresh_up=thresh_up,
          method ="dbl_sine", cutoff=cutoff, digits=digits, cumulative=cumulative, quiet=quiet)
}

#' Determine the case
#' Determine the relationship between the daily min and max and the upper & lower thresholds
#' @keywords internal

dd_case <- function(daily_min, daily_max, thresh_low, thresh_up) {

  # Six possible relationships can exist between the daily temperature cycle
  # and the upper and lower developmental thresholds. The temperature cycle can be:
    # 1) completely above both thresholds
    # 2) completely below both thresholds,
    # 3) entirely between both thresholds
    # 4) intercepted by the lower threshold
    # 5) intercepted by the upper threshold
    # 6) intercepted by both thresholds.

  if (length(daily_min) * length(daily_max) != 1) stop("dd_case() is not vectorized, please pass only one set of values at a time")

  if (daily_min >= thresh_up) {
    1

  } else if (daily_max <= thresh_low) {
    2

  } else if (daily_min >= thresh_low && daily_max <= thresh_up) {
    3

  } else if (daily_min < thresh_low && daily_max <= thresh_up) {
    4

  } else if (daily_min >= thresh_low && daily_max > thresh_up) {
    5

  } else if (daily_min < thresh_low && daily_max > thresh_up) {
    6

  } else {
    stop("Unknown case")
  }

}

#' Compute half-day GDD using triangle functions.
#' This is an internal function, called by dd_calc()
#' @keywords internal

dd_dbltri_half <- function(case, tmin, tmax, thresh_low, thresh_up) {
  switch(case,
         (thresh_up - thresh_low) / 2, # case 1
         0, # case 2
         6 * (tmax + tmin - 2 * thresh_low) / 24, # case 3
         (6 * (tmax - thresh_low)^2 / (tmax - tmin)) / 24, # case 4
         (6 * (tmax + tmin - 2 * thresh_low) / 24) - ( ( 6 * (tmax - thresh_up) ^ 2 / (tmax - tmin)) / 24), # case 5
         ((6 * (tmax - thresh_low)^2 / (tmax - tmin))  - (6 * (tmax - thresh_up)^2 / (tmax - tmin))) / 24 # case 6
  )

}

#' Compute half-day GDD using sine functions.
#' This is an internal function, called by dd_calc()
#' @keywords internal

dd_dblsine_half <- function(case, tmin, tmax, thresh_low, thresh_up) {

  alpha <- (tmax - tmin) / 2

  if (case == 1) {
    ds_dday <- (thresh_up - thresh_low) / 2

  } else if (case == 2) {
    ds_dday <- 0

  } else if (case == 3) {
    ds_dday <- 0.5 * (((tmax + tmin) / 2) - thresh_low)

  } else if (case == 4) {
    theta1 <- asin((thresh_low - (tmax + tmin) / 2) / alpha)
    ds_dday <- (1 / (2 * pi)) * ((((tmax + tmin) / 2) - thresh_low) * (pi/2 - theta1) + alpha * cos(theta1))

  } else if (case == 5) {
    theta2 <- asin((thresh_up - (tmax + tmin) / 2) / alpha)
    ds_dday <- (1 / (2 * pi)) * (((( tmax + tmin) / 2) - thresh_low )*(pi/2 + theta2) + (thresh_up - thresh_low)*(pi/2 - theta2) - (alpha * cos(theta2)) )

  } else if (case == 6) {
    theta1 <- asin((thresh_low - (tmax + tmin) / 2) / alpha)
    theta2 <- asin((thresh_up - (tmax + tmin) / 2) / alpha)
    ds_dday <- (1 / (2 * pi)) * ((((tmax + tmin) / 2) - thresh_low)*(theta2 - theta1) + alpha*(cos(theta1)-cos(theta2)) + (thresh_up - thresh_low)*(pi/2 - theta2))

  }

  ds_dday

}

